level07/src/FirstApp/AppM.hs

```diff
@@ -59,65 +51,61 @@ data Env = Env
 --     mightFail :: IO (Either Error Int)
 --     alsoMightFail :: Int -> IO (Either Error Value)
 --
--- We can wrap our functions with ExceptT and we can work directly with the
+-- We can wrap our functions with AppM and we can work directly with the
 -- values we expect to appear on the happy path, knowing that if the sad path is
--- encountered, the structure of the monad will automatically handle it for us.
--- We then `runExceptT` similar to how we `runReaderT` and the final value is an
--- `Either e a` that we can handle once, instead of at each and every place an
--- error may occur:
---
--- foo :: ExceptT Error IO Value
--- foo = do
---  a <- mightFail'
---  alsoMightFail' a
---  where
---    mightFail' :: ExceptT Error IO Int
---    mightFail' = ExceptT mightFail
---
---    alsoMightFail' :: Int -> ExceptT Error IO Value
---    alsoMightFail' = ExceptT . alsoMightFail
---
--- ( runExceptT foo ) :: IO (Either Error Value)
---
--- But how to extend our AppM ?
---
--- We could change to use the ExceptT over IO, but we would lose the
--- functionality provided by the ReaderT.
---
--- Recall that the type of the ReaderT r m a ~ r -> m a
---
--- ReaderT is a monad transformer over or for some base `m`, our initial use
--- case was to transform over `IO` as our base.
---
--- This polymorphism allows us to 'stack' our monad transformers, effectively
--- combining the functionality of the different transformer types into a single
--- monad.
---
--- But ExceptT has an `m` type variable, so what do we put there?
---
-newtype AppM a = AppM
-  { unAppM :: ReaderT Env (ExceptT Error IO) a }
-  deriving ( Functor
-           , Applicative
-           , Monad
-           , MonadIO
-           , MonadReader Env
-           -- This is a new instance we've derived because we want to be able to
-           -- `throwError` our own Error type.
-           , MonadError Error
-           )
-
--- Now that we've two transformers in our 'stack', we have to 'run' them in order
--- to effect the computations and produce our result in our base monad. When
--- you're running monad transformers you have to unpack them in order. Since our outer
--- transformer is a ReaderT, we have to run that first. Followed by running
--- the ExceptT to retrieve our `IO (Either Error a)`.
+-- encountered, the structure of our AppM will automatically handle it for us.
```

Comment:
If you're happy with the comment I added above for AppM, maybe mentioning we've added a new _effect_ here would be a good idea.

Also it'll be interesting to see how people go with implementing this from scratch compared to immediately introducing ReaderT and then ExceptT.